### PR TITLE
[Merged by Bors] - feat(RingTheory/MvPolynomial): multivariate polynomials preserve localization

### PR DESCRIPTION
**Pull request by chrisflav**
If `S` is the localization of `R` at a submonoid `M`, then `MvPolynomial σ S` is the localization of `MvPolynomial σ R` at the image of `M` in `MvPolynomial σ R`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Finsupp.lean [hunk 1]
```diff
@@ -644,6 +644,13 @@ lemma sum_cons' [AddCommMonoid M] [AddCommMonoid N] (n : ℕ) (σ : Fin n →₀
   simp_rw [Fin.sum_univ_succ, cons_zero, cons_succ]
   congr
 
+@[to_additive]
+lemma prod_mul_eq_prod_mul_of_exists [DecidableEq α] [Zero M] [CommMonoid N]
+    {f : α →₀ M} {g : α → M → N} {n₁ n₂ : N}
+    (h : ∃ a ∈ f.support, g a (f a) * n₁ = g a (f a) * n₂) :
```
Comment (erdOne) [inline Mathlib/Algebra/BigOperators/Finsupp.lean:650]: Should `h` be unbundled? (That is `(a) (ha : f.support) (ha' : g a (f a) * n₁ = g a (f a) * n₂)`)

File: Mathlib/Algebra/MvPolynomial/Basic.lean [hunk 2]
```diff
@@ -1539,7 +1539,6 @@ theorem aeval_X (s : σ) : aeval f (X s : MvPolynomial _ R) = f s :=
   eval₂_X _ _ _
 #align mv_polynomial.aeval_X MvPolynomial.aeval_X
 
-@[simp]
 theorem aeval_C (r : R) : aeval f (C r) = algebraMap R S₁ r :=
```
Comment (erdOne) [inline Mathlib/Algebra/MvPolynomial/Basic.lean:1542]: What is the reasoning behind this? These look fine to me at first glance.

File: Mathlib/RingTheory/MvPolynomial/Basic.lean [hunk 3]
```diff
@@ -169,6 +169,21 @@ instance [Finite σ] (N : ℕ) : Module.Finite R (restrictTotalDegree σ R N) :=
 
 end Degree
 
+section Algebra
+
+variable {R S σ : Type*} [CommSemiring R] [CommSemiring S] [Algebra R S]
+
+/-- If `S` is an `R`-algebra, then `MvPolynomial σ S` is a `MvPolynomial σ R` algebra. -/
+noncomputable instance : Algebra (MvPolynomial σ R) (MvPolynomial σ S) :=
```
Comment (erdOne) [inline Mathlib/RingTheory/MvPolynomial/Basic.lean:177]: This gives a diamond for `Algebra (MvPolynomial σ R) (MvPolynomial σ (MvPolynomial σ R))`. Please include the warning in the docstring and maybe not have it as a instance by default.

File: Mathlib/Algebra/MvPolynomial/Basic.lean (hunk 2 above)
Comment (chrisflav) [reply Mathlib/Algebra/MvPolynomial/Basic.lean:1542]: With the new `algHom_C`, this is provable by `simp`.

File: Mathlib/RingTheory/MvPolynomial/Basic.lean (hunk 3 above)
Comment (chrisflav) [reply Mathlib/RingTheory/MvPolynomial/Basic.lean]: I made it a `local instance`.

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Comment by mattrobball** [comment]
!bench

**Comment by mattrobball** [comment]
bors merge

**Comment by chrisflav** [comment]
Thanks for the reviews!

**Pull request closed without merging**